SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialSender.cpp

```diff
@@ -146,30 +166,82 @@ void USpatialSender::GainAuthorityThenAddComponent(USpatialActorChannel* Channel
+			NewComponentIds.Add(ComponentId);
 		}
 	});
 
-	FWorkerComponentUpdate Update = EntityACL->CreateEntityAclUpdate();
+	// If this worker is EntityACL authoritative, we can directly update the component IDs to gain authority over.
+	if (StaticComponentView->HasAuthority(EntityId, SpatialConstants::ENTITY_ACL_COMPONENT_ID))
+	{
+		const FClassInfo& ActorInfo = ClassInfoManager->GetOrCreateClassInfoByClass(Channel->Actor->GetClass());
+		const WorkerAttributeSet WorkerAttribute{ ActorInfo.WorkerType.ToString() };
+		const WorkerRequirementSet AuthoritativeWorkerRequirementSet = { WorkerAttribute };
+
+		EntityAcl* EntityACL = StaticComponentView->GetComponentData<EntityAcl>(Channel->GetEntityId());
+		for (auto& ComponentId : NewComponentIds)
+		{
+			EntityACL->ComponentWriteAcl.Add(ComponentId, AuthoritativeWorkerRequirementSet);
+		}
+
+		FWorkerComponentUpdate Update = EntityACL->CreateEntityAclUpdate();
+		Connection->SendComponentUpdate(Channel->GetEntityId(), &Update);
+	}
+
+	// Update the ComponentPresence component with the new component IDs. If this worker does not have EntityACL
+	// authority, this component is used to inform the enforcer of the component IDs to add to the EntityACL.
+	check(StaticComponentView->HasAuthority(EntityId, SpatialConstants::COMPONENT_PRESENCE_COMPONENT_ID));
+
+	ComponentPresence* ComponentPresenceData = StaticComponentView->GetComponentData<ComponentPresence>(EntityId);
+	ComponentPresenceData->AddComponentIds(NewComponentIds);
+
+	FWorkerComponentUpdate Update = ComponentPresenceData->CreateComponentPresenceUpdate();
 	Connection->SendComponentUpdate(Channel->GetEntityId(), &Update);
 }
 
-void USpatialSender::SendRemoveComponent(Worker_EntityId EntityId, const FClassInfo& Info)
+void USpatialSender::SendRemoveComponentForClassInfo(Worker_EntityId EntityId, const FClassInfo& Info)
 {
+	TArray<Worker_ComponentId> ComponentsToRemove;
 	for (Worker_ComponentId SubobjectComponentId : Info.SchemaComponents)
 	{
 		if (SubobjectComponentId != SpatialConstants::INVALID_COMPONENT_ID)
 		{
 			NetDriver->Connection->SendRemoveComponent(EntityId, SubobjectComponentId);
+			ComponentsToRemove.Add(SubobjectComponentId);
 		}
 	}
 
+	// Update ComponentPresence.
+	check(StaticComponentView->HasAuthority(EntityId, SpatialConstants::COMPONENT_PRESENCE_COMPONENT_ID));
+
+	ComponentPresence* ComponentPresenceData = StaticComponentView->GetComponentData<ComponentPresence>(EntityId);
+	const bool bRemovedComponents = ComponentPresenceData->RemoveComponentIds(ComponentsToRemove);
+	if (bRemovedComponents)
+	{
+		FWorkerComponentUpdate Update = ComponentPresenceData->CreateComponentPresenceUpdate();
+		Connection->SendComponentUpdate(EntityId, &Update);
+	}
+
 	PackageMap->RemoveSubobject(FUnrealObjectRef(EntityId, Info.SchemaComponents[SCHEMA_Data]));
 }
 
+void USpatialSender::SendRemoveComponentForComponentId(Worker_EntityId EntityId, Worker_ComponentId ComponentId)
```

Comment:
Nit - the name seems more complicated than it needs to be. It seems more like a TrySendRemoveComponent.

Reply:
Agree it's complicated, but not sure "TrySend" is apt? It should always succeed. It's basically SendComponentUpdateWiithPresenceUpdate, which is probably more verbose than what I had before? If you think TrySendRemoveComponnent is best I'll go with that?